### ISSUE TITLE
focusIn listener causes issues with outside elements

### ISSUE THREAD

**Issue by markhalliwell** [focus-trap/focus-trap#806]
Re: #42 

When certain Kendo UI elements are attached to the body outside the focus-trap (i.e. [datepicker](https://www.telerik.com/kendo-vue-ui/components/dateinputs/datepicker/)), it attempts to regain focus in the `focusIn` listener; even when using `allowOutsideClick` or `clickOutsideDeactivates`.

Not sure what the fix needs to be, but from the little time I have spent with this library, I think checking for these options in the `focusIn` handler is also needed.

Optionally, a new option could be introduced that either takes a string (selector) or a callback that can check if the clicked/focused element is allowed to take focus.

**Comment by far-fetched** [comment]
:wave: Hello :wave: 

I had a similar problem. I figured out that at the moment when I attach a new DOM element (open date picker in your case) I deactivated the existing focus trap and activated a new, thus old focus trap returned focus to the element which activated the trap. So my solution was to add `returnFocusOnDeactivate: false` as option. Maybe that is also your case. If not, Can you provide a minimal working example, we will try to recognize the issue.
thanks !

**Comment by stefcameron** [comment]
@far-fetched Thank you for chiming in!

@markhalliwell Are the elements always added directly to the body (and so outside any possible trap containers) and positioned at the right place using CSS, or is it possible to tell KendoUI where to place them in the DOM (ie what their parent should be)? If the latter is possible, can you have it placed inside one of your trap's containers?

**Comment by markhalliwell** [comment]
> I figured out that at the moment when I attach a new DOM element (open date picker in your case) I deactivated the existing focus trap and activated a new, thus old focus trap returned focus to the element which activated the trap.

The issue here is that the DatePicker is a component nested inside a slot of a custom Dialog component (wrapper around Kendo UI's); it has no reference of the parent Dialog where the focus trap is. Also, it would require event handlers to be placed on all Kendo UI components that do this, which adds tech debt and code parity issues.

> So my solution was to add `returnFocusOnDeactivate: false` as option.

Unfortunately, due to keyboard power users, this is not an option. When the Dialog closes, the focus must be returned to the component (button/link) that opened it.

> Can you provide a minimal working example

Sure! https://codesandbox.io/s/hardcore-joji-zm0ghh


> Are the elements always added directly to the body (and so outside any possible trap containers) and positioned at the right place using CSS

Yes. By default, the Kendo UI DatePicker always attaches itself to the `BODY` element (as with many other similar components).

> is it possible to tell KendoUI where to place them in the DOM (ie what their parent should be)? If the latter is possible, can you have it placed inside one of your trap's containers?

Technically, it is possible via [the `popupSettings` prop](https://www.telerik.com/kendo-vue-ui/components/dateinputs/api/DatePickerProps/#toc-popupsettings). However, the DatePicker cannot be inside the Dialog because it scrolls any overflow; thus hiding the DatePicker if it's attached inside the dialog where the focus-trap is. This is true of many dynamic UI elements that require positioning its element near or close to the container of the dialog; this is why they're attached to the `BODY` in the first place.

**Comment by markhalliwell** [comment]
> You mentioned the term `slot` here, so does that mean your gets placed in a web component's (i.e. shadow DOM's) `<slot>`?

I was referring to [Vue.js Slots](https://vuejs.org/guide/components/slots.html). It's a common practice/pattern for wrapping content in Vue. Essentially, it's just a way to inject content into a component; it could have easily been called `<content />`, but they opted for `<slot />` instead; IDK, it is confusing at times.

From the codesandbox example above, the main contents of the dialog (including the DatePicker) are [in the `App.vue` component](https://codesandbox.io/s/hardcore-joji-zm0ghh?file=/src/App.vue:518-1926) while the `Dialog.vue` component just wraps the provided content ([denoted using `<slot />`](https://codesandbox.io/s/hardcore-joji-zm0ghh?file=/src/Dialog.vue:69-82)) with additional structure relative to the component; in this case, a [Kendo UI Dialog](https://codesandbox.io/s/hardcore-joji-zm0ghh?file=/src/Dialog.vue:11-69).

This allows us to effectively create a single Dialog component that has additional features ([focus-trap for example](https://codesandbox.io/s/hardcore-joji-zm0ghh?file=/src/Dialog.vue:528-1061)) that can be reused across the codebase. It also allows the contents of the Dialog to live with the implementation which also provides context. If we didn't follow this pattern, we'd have to implement a custom Dialog for each implementation thus duplicating a bunch of code.


> If so, have you tried using the [getShadowRoot](https://github.com/focus-trap/tabbable#getshadowroot-option) tabbable option (use focus-trap's `tabbableOptions` to specify it)?

No; not applicable

**Comment by dstoch** [comment]
I have the same problem with UI elements which are attached to body eg.: Select2, dropdown panels which contain some interactive elements (eg. links). They cannot be placed inside trap container in modal, because then they could be cut off by some overflow settings, like @markhalliwell wrote earlier. I think many libraries do it this way.

Currently I am investigating possible solutions for this problem...

**Comment by stefcameron** [comment]
@markhalliwell 
> Technically, it is possible via [the `popupSettings` prop](https://www.telerik.com/kendo-vue-ui/components/dateinputs/api/DatePickerProps/#toc-popupsettings). However, the DatePicker cannot be inside the Dialog because it scrolls any overflow; thus hiding the DatePicker if it's attached inside the dialog where the focus-trap is. This is true of many dynamic UI elements that require positioning its element near or close to the container of the dialog; this is why they're attached to the `BODY` in the first place.

I was thinking, if you can technically attach them to a designated parent, could you create a little `<div>` off the body into which they could be placed, and then maybe including this div as an additional container in focus trap should mean that anything that gets added to that container would be considered part of the trap.

**Comment by markhalliwell** [comment]
The reason I phrased it that way was because it would be a lot of effort to try and "work around" the issue and I would have to implement the same hack across all the code; for each component.

I don't understand the hesitation to add additional check in the `focusIn` handler to ensure it doesn't steal focus from a valid outside click.

**Comment by stefcameron** [comment]
I haven't thought too deeply yet about your suggestion, but on the surface, focus trap is meant to work with containers in which focus is trapped now you basically want to say that focus can escape the trap... That's sounding like a lot of overhead on the focus trap side to make that work, and sounds like it would break a lot of assumptions. Might also break tab order, which we try to delegate to the browser as much as possible. And if we can find a reasonable work around in your code, I'd rather that than immediately jumping and opening Pandora's box. 

From my side, I don't see why you can't apply the same config across components. Doesn't seem like that much, but I don't use Kendo, so I could be mistaken there.

**Comment by markhalliwell** [comment]
> focus trap is meant to work with containers in which focus is trapped now you basically want to say that focus can escape the trap...

Yes, most dynamic UI components that are "popup" in nature attach themselves to the `BODY` element. This is a **widely used** and **common practice** amongst front-end frameworks; specifically designed to address the many issues that arise when nesting "popup" components (i.e. hidden overflow, style inheritance/overrides, event bubbling, etc.).

> And if we can find a reasonable work around in your code, I'd rather that than immediately jumping and opening Pandora's box.

Except that `clickOutsideDeactivates` and `allowOutsideClick` already opened this "Pandora's box". These are options that already exist and are supposedly supported by this project. I created this issue to bring light to a **bug** with these options where `checkFocusIn` does not respect these options and steals focus back when it shouldn't.

> That's sounding like a lot of overhead on the focus trap side to make that work, and sounds like it would break a lot of assumptions

Nope. It's literally just adding a few lines of code to check for the existing options inside [`checkFocusIn`](https://github.com/focus-trap/focus-trap/blob/master/index.js#L407-L421).

Similar to how [`checkClick`](https://github.com/focus-trap/focus-trap/blob/master/index.js#L563-L569) does it.

> From my side, I don't see why you can't apply the same config across components. Doesn't seem like that much, but I don't use Kendo, so I could be mistaken there.

I'm not trying to sound ungrateful. This package does indeed do 99% of what we need it to do. However, to simply dismiss a legitimate use case just because is "seems" counter-productive to it achieving its primary goal is not a very strong argument.

I'm assuming that is why these options were created in the first place: to give the developer (who has to handle these use cases) the _ability_ to do so. It's up to them to determine when this option evaluates to true. In our case: it's when the outside component is a Kendo UI "popup" component; for others, it may be a Select2 instance.

I'm not asking for anything that wasn't already added. I'm asking for a bug with those options to be fixed.

Believe me, if I could have opted for a "quick and easy" solution, I would have. I really do not enjoy filing issues like this. It usually takes a ton of back and forth for me to try and "convince" someone of a legitimate issue like this.

**Comment by stefcameron** [comment]
I understand you're frustrated, but I didn't make this library from scratch. I agreed to take it over (and tabbable, and focus-trap-react) for the benefit of everyone relying on it, and to try to maintain it and improve it. So I don't know all the history. I'm trying to bring them all forward as best I can, and without making a mess (and IMO, these libraries have been a mess for a while; so many options tacked on over the years at every whim). Why what was done, I don't have the full history or context.

> Except that clickOutsideDeactivates and allowOutsideClick already opened this "Pandora's box". 

IIRC,

- `clickOutsideDeactivates`: Not a good comparison here. This doesn't "add" an element into the trap that isn't in one of the containers already. It gives you a chance to decide if the trap will deactivate on a click outside all containers. If you say no, that thing you clicked on doesn't get focus. The trap will bring focus back into one of the containers.

- `allowOutsideClick`: Same thing. Not a good comparison, and for basically the same reasons. While it lets the click through, I don't believe it ever lets the focus out of the containers. And this option is for clicks on elements outside all configured containers. You can decide if the click will go through, but the trap continues to trap the focus. That thing you allowed clicking on never actually gets focus.

So I think what you're asking for here is actually quite different from how the trap operates today, and I'm hesitant to believe that it's "just 3 lines". Oh, how many times I have convinced myself of that very thing, "it's just a few lines", only to find I've spent "just a few days" working on it.

You might see this as a little bug; I'm not there yet. Seems like quite a change to me, and a change contrary to how the trap is designed to work today.

To do what you want, it seems we'd have to add random nodes to what we consider the trap. Maybe we generate some type of fake container internally, because this would have to keep working with `updateContainers()` (or whatever the name of that function is; I don't recall right this minute and can't lookup). And where would they go in terms of tab order? And what about tab order between them, because surely you'll want to add more than just one (or someone will)? Focus trap isn't in the business of determining tab order. It stays out of it as much as it can, but I think this kind of thing would force it to get much more involved, and then it'll become subjective as to what the order should be.

**Comment by dstoch** [comment]
Maybe my use case is not exatly the same as Mark's, but in general when we are talking about elements added directly to BODY I see two main problems (you've described it also eralier):
1. We cannot control 3rd party components to force them to add elements to some concrete container below body. They often add elements to body tag directly. So we cannot pass specific element below body when constructing focus-trap, we must update it using updateContainers() on the fly. I've already implemented this for Select2 using MutationObserver to observe changes in document.body and dynamically update list fo containers in focus-trap. Here possible improvement could be to allow to pass some "dynamic" selector to focus-trap, not concrete elements. But it is probably quite big change in code.
2. Event if we solved this first problem, the second one still exists: proper tab order (as @stefcameron  decribed in last comment). And currently I don't know how to solve this.

The whole problem comes form HTML/CSS limitations to handle such things like modals, dropdowns, etc. These all hacks with adding elements to body are problematic for operating from the keyboard becuase this braeks natural HTML elements order. Maybe someone knows some applications/websites when these things are properly implemented (eg. some government sites, which often are WCAG compliant)?
